### PR TITLE
[15950] Tutorial - External Locators

### PR DESCRIPTION
**Pull request by Mario-DL**
This PR adds a tutorial for leveraging Fast DDS' external locators feature in Vulcanexus.

Cannot be merged before:

*  eprosima/Fast-DDS#3079
* #86

### CODE REVIEW

**Review by EduPonz — CHANGES_REQUESTED** [review]

File: docs/rst/tutorials/core/deployment/external_locators/external_locators.rst [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+                                </initialPeersList>
+                            </builtin>
+                        </rtps>
+                    </participant>
+                </profiles>
+            </dds>
+
+.. note::
+
+    Note that the docker network itself does not create another level of externality in this case, as the containers are bridged with the host network.
```
[the remark this answers is not in the document]
Comment (Mario-DL) [reply docs/rst/tutorials/core/deployment/external_locators/external_locators.rst:187]: I moved it before the XML file description, after the line referring to the example above

`Following the example above ....
.. note:: 
....`

File: docs/rst/tutorials/core/deployment/external_locators/external_locators.rst [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+                                        <metatraffic_external_unicast_locators>
+                                            <udpv4 externality="1" cost="0" mask="24">
+                                                <!-- Host 1 external IP -->
+                                                <address>192.168.1.40</address>
+                                                <port>11200</port>
+                                            </udpv4>
+                                        </metatraffic_external_unicast_locators>
+                                        <!-- Locators of remote participants (discovery traffic)-->
+                                        <initialPeersList>
+                                            <!--container 1 peer-->
```
Comment (EduPonz) [inline docs/rst/tutorials/core/deployment/external_locators/external_locators.rst:98]:
```suggestion
                                            <!-- Container 1 peer-->
```

File: docs/rst/tutorials/core/deployment/external_locators/external_locators.rst [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+                                            <locator>
+                                                <udpv4>
+                                                    <!-- Host 2 external IP -->
+                                                    <address>192.168.1.56</address>
+                                                    <port>11200</port>
+                                                </udpv4>
+                                            </locator>
+                                            <!-- local network multicast. Discover
+                                            other participants in the same LAN,
+                                            using External Locators, or not -->
```
Comment (EduPonz) [inline docs/rst/tutorials/core/deployment/external_locators/external_locators.rst:108]:
```suggestion
                                            <!--
                                                Local network DDS default multicast to discover
                                                other participants in the same LAN,
                                                using External Locators, or not
                                            -->
```

File: docs/rst/tutorials/core/deployment/external_locators/external_locators.rst [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+
+.. tabs::
+
+    .. tab:: HOST 1
+
+        .. tabs::
+
+            .. tab:: CONTAINER 1
+
+                .. code-block:: xml
```
Comment (EduPonz) [inline docs/rst/tutorials/core/deployment/external_locators/external_locators.rst:72]: I'd move all the XML files to several resource files (like in #86) and include them with literal include.

File: docs/rst/tutorials/core/deployment/external_locators/external_locators.rst [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+                                            <locator>
+                                                <udpv4>
+                                                    <!-- Host 1 external IP -->
+                                                    <address>192.168.1.40</address>
+                                                    <port>11200</port>
+                                                </udpv4>
+                                            </locator>
+                                            <!-- local network multicast. Discover
+                                            other participants in the same LAN,
+                                            using External Locators, or not -->
```
Comment (EduPonz) [inline docs/rst/tutorials/core/deployment/external_locators/external_locators.rst:164]: Same as above

File: docs/rst/tutorials/core/deployment/external_locators/external_locators.rst [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+                                </rtps>
+                            </participant>
+                        </profiles>
+                    </dds>
+
+
+Run the example
+----------------
+
+This section provides with step-by-step instructions for setting up the example scenario described in the previous Background section.
```
Comment (EduPonz) [inline docs/rst/tutorials/core/deployment/external_locators/external_locators.rst:182]:
```suggestion
This section provides with step-by-step instructions for setting up the example scenario described in previous sections.
```

File: docs/rst/tutorials/core/deployment/external_locators/external_locators.rst [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+
+
+.. note::
+
+    It is important to specify the port mapping argument so as to expose docker internal ports to the host. See `Docker Networking <https://docs.docker.com/config/containers/container-networking/>`_ for further information.
+
+The next step is the creation of the XML profiles.
+Inside each one of the three containers, create a Profiles.xml file and paste the contents of the corresponding XML profile configuration, according to the previous section.
+
+Finally, export the environment variable pointing to the Profiles file, source Vulcanexus environment and run the ros2 example nodes.
```
Comment (EduPonz) [inline docs/rst/tutorials/core/deployment/external_locators/external_locators.rst:235]:
```suggestion
Finally, export the environment variable pointing to the Profiles.xml file, source Vulcanexus environment and run the ros2 example nodes.
```

File: docs/rst/tutorials/core/deployment/external_locators/external_locators.rst [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+    .. tab:: HOST 1
+
+        .. tabs::
+
+            .. tab:: CONTAINER 1
+
+                .. code-block:: bash
+
+                    source vulcanexus_entrypoint.sh
+                    export FASTRTPS_DEFAULT_PROFILES_FILE=/Profiles.xml #Or the Profiles.xml file location
```
Comment (EduPonz) [inline docs/rst/tutorials/core/deployment/external_locators/external_locators.rst:248]:
```suggestion
                    export FASTRTPS_DEFAULT_PROFILES_FILE=/Profiles.xml # Or the Profiles.xml file location
```

File: docs/rst/tutorials/core/deployment/external_locators/external_locators.rst [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+    .. tab:: HOST 2
+
+        .. tabs::
+
+            .. tab:: CONTAINER 1
+
+                    .. code-block:: bash
+
+                        source vulcanexus_entrypoint.sh
+                        export FASTRTPS_DEFAULT_PROFILES_FILE=/Profiles.xml #Or the Profiles.xml file location
```
Comment (EduPonz) [inline docs/rst/tutorials/core/deployment/external_locators/external_locators.rst:268]:
```suggestion
                        export FASTRTPS_DEFAULT_PROFILES_FILE=/Profiles.xml # Or the Profiles.xml file location
```

File: docs/rst/tutorials/core/deployment/external_locators/external_locators.rst [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+                        ros2 run demo_nodes_cpp listener
+
+
+At this point, nodes should be communicating with each other as expected.
+A message `Hellow World: [count]` should start printing in the talker's container terminal while both listeners keep receiving it, in their respective container consoles, as follows:
```
Comment (EduPonz) [inline docs/rst/tutorials/core/deployment/external_locators/external_locators.rst:273]:
```suggestion
A message `Hello World: [count]` should start printing in the talker's container terminal while both listeners keep receiving it, in their respective container consoles, as follows:
```

File: docs/rst/tutorials/core/deployment/external_locators/external_locators.rst [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+
+.. raw:: html
+
+    <video width=100% height=auto autoplay loop controls>
+        <source src="../../../../../_static/resources/tutorials/core/deployment/external_locators/external_locators.mp4">
```
Comment (EduPonz) [inline docs/rst/tutorials/core/deployment/external_locators/external_locators.rst:278]: Nitpick, the host 2 listener seems to be receiving data before the host 1 talker sends it

**Review by EduPonz — CHANGES_REQUESTED** [review]
I'd like to congratulate you on the video!

**Comment by EduPonz** [comment]
I've rebased this on top of #86 again since there I had to add the video as raw HTML instead of through the extension because that was breaking the latex compilation on RTD. I've enabled a [build on this branch](https://docs.vulcanexus.org/en/docs-ext_loc_tutorial/rst/tutorials/core/deployment/external_locators/external_locators.html) so we can see how it looks online before merging.

File: docs/rst/tutorials/core/deployment/external_locators/external_locators.rst [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+                        export FASTRTPS_DEFAULT_PROFILES_FILE=/Profiles.xml #Or the Profiles.xml file location
+                        ros2 run demo_nodes_cpp listener
+
+
+At this point, nodes should be communicating with each other as expected.
+A message `Hellow World: [count]` should start printing in the talker's container terminal while both listeners keep receiving it, in their respective container consoles, as follows:
+
+.. raw:: html
+
+    <video width=100% height=auto autoplay loop controls>
```
Comment (EduPonz) [inline docs/rst/tutorials/core/deployment/external_locators/external_locators.rst:277]:
```suggestion
    <video width=100% height=auto autoplay loop controls muted>
```

**Review by EduPonz — CHANGES_REQUESTED** [review]

File: docs/rst/tutorials/core/deployment/external_locators/external_locators.rst [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+
+Many robotic applications involve complex network topologies (mesh, nested, etc.) in which nodes, processes or entities from one host may need to interact with both, nodes from outside and inside its network.
+
+Consider the following nested network scenario: one talker node inside one host's private LAN sub-network needs to communicate with two listener nodes; the first one belonging to the same LAN as the talker, and the second one deployed within an external host with its own private sub-network.
+Also consider both hosts externally connected to the same network.
+The following diagram depicts the aforementioned scheme.
+
+.. figure:: /rst/figures/tutorials/core/external_locators/network_setup.svg
+   :align: center
+   :scale: 75%
```
[the remark this answers is not in the document]
Comment (Mario-DL) [reply docs/rst/tutorials/core/deployment/external_locators/external_locators.rst:22]: Thanks !

**Review by EduPonz — APPROVED** [review]
Tentatively approving as the docs look very nice to me. However, I'll not merge it until I can run the tutorial once the following is merged:

*  eprosima/Fast-DDS#3079

**Review by EduPonz — APPROVED** [review]

**Pull request merged**